Federation Setup Console.

ferdeen opened the pull request:
@bokobza, I've added GenesisMiner and MultisigAddressCreator to console app.

![image](https://user-images.githubusercontent.com/16000683/48914045-32c12000-ee72-11e8-923d-099059a1ccec.png)

@monsieurleberre, do we need to include the creation of public private keys (-p) - I kept the code in there as wasn't too sure.  

Fixes #190.

Cheers.

ferdeen commented on src/FedKeyPairGen/Program.cs:
```diff
@@ -27,37 +30,53 @@ static void Main(string[] args)
+                if (args.Contains("-p"))
+                {
+                    var mnemonic = new Mnemonic(Wordlist.English, WordCount.Twelve);
+                    var pubKey = mnemonic.DeriveExtKey().PrivateKey.PubKey;
```
@monsieurleberre do we still need this as part of the setup ?

bokobza commented:
A few points:
* Is the command `fedsetup` or `federationsetup`?
* The verbs for the commands should not be `Creates`, but `Create` as it's an action that's described.
* `-a` for the multisig could be `-m` instead.
* the creation of multisig is missing the arguments. You need to have a number and some public keys. For example if you want to have a 2 of 3 multisig you need to pass 2 and then 3 pub keys.
* -p the instructions could be clearer here. Also how many are you generating? 1 or 2?

ferdeen commented:
@bokobza 

> 
> 
> A few points:
> 
>     * Is the command `fedsetup` or `federationsetup`?
> 
>     * The verbs for the commands should not be `Creates`, but `Create` as it's an action that's described.
> 
>     * `-a` for the multisig could be `-m` instead.
> 
>     * the creation of multisig is missing the arguments. You need to have a number and some public keys. For example if you want to have a 2 of 3 multisig you need to pass 2 and then 3 pub keys.
> 
>     * -p the instructions could be clearer here. Also how many are you generating? 1 or 2?

- federationsetup - fedsetup wasn't clear - I've replaced any reference to fedsetup in the code.
- Re verbs okay I've changed
- Changed -a switch to -m
- In master the output usage was `fedkeypairgen [-name=<name>] [-folder=<output_folder>] [-pass=<password>] [-h]` - this wasn't fully implemented.  I'll implement it now.
- The the above is implemented the I'll make the -p switch more clear.

Cheers,

Ferdeen

monsieurleberre commented:
Maybe somethings to think about:

I have used that package before to handle complex commandline inputs, it helps a lot validating arguments and generating help (up to you):
https://github.com/commandlineparser/commandline

The number of public keys should be enough to know the size of the federation (so no need to pass an N parameter). However it is assumed that the number is odd, and no more than 15 in our spec, maybe a validation for that. Also the quorum for the multisig (M parameter) needs to be >N/2

ferdeen commented:
Applied changes.  Points of interest :

Run each option in debug.
![image](https://user-images.githubusercontent.com/16000683/48944178-95bbc100-ef1d-11e8-955e-a3d169c22b0c.png)

Output for Help.
![image](https://user-images.githubusercontent.com/16000683/48944234-cd2a6d80-ef1d-11e8-91e9-fa2985f24216.png)

Output for Creating a genesis block.
![image](https://user-images.githubusercontent.com/16000683/48944260-ec28ff80-ef1d-11e8-9f14-85b477ce2cda.png)

Output for creating public private keys.
![image](https://user-images.githubusercontent.com/16000683/48944253-e3382e00-ef1d-11e8-85e6-4b1cd8690b7c.png)

Output for creating multisig addresses
![image](https://user-images.githubusercontent.com/16000683/48944273-f77c2b00-ef1d-11e8-8109-e2a9be3be738.png)

Let me know if we have everything covered. 

Cheers.

ferdeen commented:
@monsieurleberre just read you point above - let me apply it.  Cheers.

bokobza commented:
Output for Creating a genesis block: OK
Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?
Output for creating multisig addresses: I see here you're creating many multisig addresses. this was there because this method was used to create all the keys in a test network. 
All you need here is the sidechain multisig address and the Mainchain multisig address. 
![image](https://user-images.githubusercontent.com/1867877/48946150-af143b80-ef24-11e8-8521-53f15e6c44b7.png)
I also see here that they're the same so maybe there is a problem in the networks used to generate each.

ferdeen commented:
> 
> 
> Output for Creating a genesis block: OK
> Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?
> Output for creating multisig addresses: I see here you're creating many multisig addresses. this was there because this method was used to create all the keys in a test network.
> All you need here is the sidechain multisig address and the Mainchain multisig address.

Fixed the above. 

With regards to :

> ![image](https://user-images.githubusercontent.com/1867877/48946150-af143b80-ef24-11e8-8521-53f15e6c44b7.png)
> I also see here that they're the same so maybe there is a problem in the networks used to generate each.

@monsieurleberre  I went through this with @bokobza we noticed that the SCRIPT_ADDRESS for all networks are set to 196.  This is why the main and side chain multisig addresses are the same.  I agree with @bokobza it would make sense for these to be different, so it will be easy to identify which multsig address belongs where.  I haven't change the value in the above commit but will probably need to look what [decimal prefix](https://en.bitcoin.it/wiki/List_of_address_prefixes) to use for each network.

![image](https://user-images.githubusercontent.com/16000683/49020563-aa4fc180-f188-11e8-805a-e44048828086.png)

With regards to :  
> Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?

@monsieurleberre Does the POA feature create a file which contains the federated members public private key for mining ? If so then the current logic I've updated in the console app may be redundant.

monsieurleberre commented:
as discussed on slack, new Prefixes are coming to avoid confusing networks
also I am not sure about the keys used to mine vs keys used to sign from multisig, I think they could be the same...

ferdeen commented:
Okay we can check during testing. Cheers.
